azure/services/securitygroups/securitygroups.go

```diff
@@ -52,111 +54,72 @@ func New(scope NSGScope) *Service {
 	}
 }
 
-// Reconcile gets/creates/updates a network security group.
+// Reconcile gets/creates/updates network security groups.
 func (s *Service) Reconcile(ctx context.Context) error {
 	ctx, span := tele.Tracer().Start(ctx, "securitygroups.Service.Reconcile")
 	defer span.End()
 
-	if !s.Scope.IsVnetManaged() {
-		s.Scope.V(4).Info("Skipping network security group reconcile in custom VNet mode")
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
+	// Only create the NSGs if their lifecycle is managed by this controller.
+	// NSGs are managed if and only if the vnet is managed.
+	managed, err := s.Scope.IsVnetManaged()
+	if err != nil {
+		return errors.Wrap(err, "failed to determine if network security groups are managed")
+	} else if !managed {
+		s.Scope.V(4).Info("Skipping network security groups reconcile in custom VNet mode")
 		return nil
 	}
 
+	// We go through the list of NSGSpecs to reconcile each one, independently of the result of the previous one.
+	// If multiple erros occur, we return the most pressing one
+	// order of precedence is: error creating -> creating in progress -> created (no error)
+	var result error
```

Comment:
I just noticed this comment actually 🙈 , with respect to https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1684#issuecomment-931091181 . I guess the question still remains as to - When we have multiple errors in the category of `error creating`, should we think about showing them all? I understand it can be too much errors, but I don't think we can find the most pressing error. In the current case we just return the last error that occurs, which may or not be the most pressing errors. And operation-not-done is an error in code, but in terms of the whole operation, it's only a stage of creation, and same for created (no error). So I think there's just mainly one category of errors (excluding operation-not-done errors) which is error in creating the resource, in this case the NSG